src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/HTTPOpenIdAuthenticator.java

```diff
@@ -0,0 +1,328 @@
+     * 1. Validate the OP is correct via TLS certificate check
+     * 2. Validate response is OK
+     * 3. Validate application type is either JSON or JWT
+     * 4. Validate response is one of: just signed, just encrypted, or signed then encrypted (encryption MUST NOT occur before signing)
+     * 5. If response is signed then validate the signature via JWS and confirm the response has "iss" and "aud" claims
+     * 6. Validate "iss" claim is equal to the issuer ID url
+     * 7. Validate the "aud" claim is equal to the client ID
+     * 8. If the client provides a userinfo_encrypted_response_alg value decrypt the response using the keys from registration
+     * 9. Validate "sub" claim is always present
+     * 10. Validate "sub" claim matches the ID token
+     * @param request The SecurityRequest to perform auth on
+     * @param context The active thread context
+     * @return AuthCredentials formed through querying the userinfo_endpoint
+     * @throws OpenSearchSecurityException On failure to extract credentials from the request
+     */
+    public AuthCredentials extractCredentials0(SecurityRequest request, ThreadContext context) throws OpenSearchSecurityException {
+
+        try (CloseableHttpClient httpClient = createHttpClient()) {
+
+            HttpGet httpGet = new HttpGet(this.userinfo_endpoint);
+
+            RequestConfig requestConfig = RequestConfig.custom()
+                .setConnectionRequestTimeout(requestTimeoutMs, TimeUnit.MILLISECONDS)
+                .setConnectTimeout(requestTimeoutMs, TimeUnit.MILLISECONDS)
+                .build();
+
+            httpGet.setConfig(requestConfig);
+            httpGet.addHeader(AUTHORIZATION_HEADER, request.getHeaders().get(AUTHORIZATION_HEADER));
+
+            // HTTPGet should internally verify the appropriate TLS cert.
+            try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
+
+                if (response.getCode() < 200 || response.getCode() >= 300) {
+                    throw new AuthenticatorUnavailableException(
+                        "Error while getting " + this.userinfo_endpoint + ": " + response.getReasonPhrase()
+                    );
+                }
+
+                HttpEntity httpEntity = response.getEntity();
+
+                if (httpEntity == null) {
+                    throw new AuthenticatorUnavailableException(
+                        "Error while getting " + this.userinfo_endpoint + ": Empty response entity"
+                    );
+                }
+
+                String contentType = httpEntity.getContentType();
+                if (!contentType.contains(APPLICATION_JSON) && !contentType.contains(APPLICATION_JWT)) {
+                    throw new AuthenticatorUnavailableException(
+                        "Error while getting " + this.userinfo_endpoint + ": Invalid content type in response"
+                    );
+                }
+
+                String userinfoContent;
+
+                try (
```

Comment:
Would it be possible to use nimbus' [`UserInfoResponse.parse(response)`](https://www.javadoc.io/doc/com.nimbusds/oauth2-oidc-sdk/5.22/com/nimbusds/openid/connect/sdk/UserInfoResponse.html)?

The KeySetRetriever use's nimbuses similarly like this: https://github.com/opensearch-project/security/blob/main/src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/KeySetRetriever.java#L92-L97

Reply:
Hmm let me look. I just googled how to read the entity content properly but if we already have something that does that, then that would be great.

Reply:
So, I took a look and I think trying to use the package is more hassle than its worth. For starters, we need to add a dependency on https://mvnrepository.com/artifact/com.nimbusds/oauth2-oidc-sdk. But more notably, trying to use the Parse logic will require resolving a translation between the CloseableHttpResponse we current use and the generic HttpResponse that the built in method expects. 

It seems like we are probably better off leaving things as they are for now.